quickwit/quickwit-indexing/src/split_store/indexing_split_cache.rs

```diff
@@ -138,30 +106,119 @@ fn split_id_from_split_folder(dir_path: &Path) -> Option<&str> {
 }
 
-pub struct LocalSplitStore {
-    inner: Mutex<InnerLocalSplitStore>,
+/// The `IndexingSplitCache` is a local cache used to improve the performance of indexing nodes.
+/// Its purpose is simple: when a new split is freshly created, it is usely merged
+/// very rapidly after.
+///
+/// In order to prevent this merge to force its download, we store it in the
+/// `IndexingSplitCache`. This store is just a cache: a cache miss is acceptable and
+/// just means that the split will be redownloaded.
+///
+/// The indexing split cache eviction policy however, is rather uncommon.
+/// On our happy path, a split is stored into the cache, and is then used only once
+/// to undergo a merge.
+///
+/// For this reason, we simply offer a way to `move splits into the cache`,
+/// and `move splits out of the cache`. A split is removed from the split store
+/// after its first access.
+///
+/// Of course a failed merge could require accessing a given split more than once. In that
+/// case the split will be downloaded again.
+///
+/// The cache size is limited by 3 things:
+/// - a maximum number of splits as defined in the `SplitStoreQuota`.
+/// - a maximum number of bytes as defined in the `SplitStoreQuota`.
+/// - finally, we evict older splits to make sure that the newest split and the oldest
+/// split only differ by at most `SPLIT_MAX_AGE`.
+///
+/// The point of this final rule invariant is to make sure that the disk space will be
+/// if the cache is NOT under pressure but some splits are actually useless.
+///
+/// When adding a new split into the cache, if adding the split would break one of the following
+/// limit, we simply remove split one by one starting by the oldest first, until the split
+/// can be added.
+
+pub struct IndexingSplitCache {
+    inner: Mutex<InnerSplitCache>,
+}
+
+struct InnerSplitCache {
+    split_registry: SplitFolderRegistry,
+    split_store_folder: PathBuf,
 }
 
-struct InnerLocalSplitStore {
+struct SplitFolderRegistry {
     /// Splits owned by the local split store, which reside in the split_store_folder.
-    /// SplitId -> SplitFolder
-    split_folders: HashMap<Ulid, SplitFolder>,
-    /// Binary heap of split ids.
-    /// The binary heap is used for eviction.
     ///
     /// Splits ids are generated using ULID, so that they are sorted
     /// according to their creation date.
+    ///
     /// We evict the oldest split first. (Note this is not an LRU strategy
     /// because we do not care about the last access time, but we only
     /// consider the creation time.)
-    split_ids: BinaryHeap<Reverse<Ulid>>,
-    /// The root folder where all data is moved into.
-    split_store_folder: PathBuf,
-    /// The split store space quota shared among all indexing split stores.
-    split_store_space_quota: SplitStoreQuota,
+    split_folders: BTreeMap<Ulid, ByteSize>,
+    /// The split store quota shared among all indexing split stores.
+    split_store_quota: SplitStoreQuota,
```

Comment:
I moved the non io stuff to this `struct` to make it easier to proofread that we cannot reach an inconsistent state.